File: android/src/main/java/app/plink/android/pairing/PairingConsent.kt
package app.plink.android.pairing

import java.security.MessageDigest
import java.util.Base64
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

/** A key preview never authorizes notification forwarding or persistence of trust. */
@Serializable
data class PairingConsent(
    val version: Int = 1,
    val stage: String,
    val confirmation: PairingConfirmation,
    val proof: String
) {
    fun verified(sessionKey: ByteArray): Boolean = runCatching {
        require(version == 1 && stage in setOf("preview", "confirmed") && sessionKey.size == 32)
        MessageDigest.isEqual(Base64.getDecoder().decode(proof), signature(stage, confirmation, sessionKey))
    }.getOrDefault(false)

    fun encode(): String = json.encodeToString(this).also { require(it.toByteArray().size <= 16_384) }

    companion object {
        private val json = Json { encodeDefaults = true; ignoreUnknownKeys = false }
        fun create(stage: String, confirmation: PairingConfirmation, sessionKey: ByteArray): PairingConsent {
            require(stage in setOf("preview", "confirmed") && sessionKey.size == 32)
            return PairingConsent(stage = stage, confirmation = confirmation, proof = Base64.getEncoder().encodeToString(signature(stage, confirmation, sessionKey)))
        }
        fun decode(payload: String): PairingConsent {
            require(payload.toByteArray().size <= 16_384 && payload.trimStart().startsWith("{"))
            return json.decodeFromString<PairingConsent>(payload).also {
                require(it.version == 1 && it.stage in setOf("preview", "confirmed"))
            }
        }
        private fun signature(stage: String, value: PairingConfirmation, key: ByteArray): ByteArray {
            val fields = listOf("plink-consent-v1", stage, value.deviceId, value.deviceName, value.platform,
                value.endpoint, value.publicKey, value.targetDeviceId, value.offerNonce, value.sessionId,
                value.protocolVersion.toString())
            val input = fields.joinToString("") { "${it.toByteArray(Charsets.UTF_8).size}:$it" }
            return Mac.getInstance("HmacSHA256").run {
                init(SecretKeySpec(key, "HmacSHA256"))
                doFinal(input.toByteArray(Charsets.UTF_8))
            }
        }
    }
}

File: macos/Sources/PlinkCore/PairingConsent.swift
import CryptoKit
import Foundation

/// A preview exchanges keys; only a confirmed message records the phone user's consent.
public struct PairingConsent: Codable, Equatable, Sendable {
    public enum Stage: String, Codable, Sendable { case preview, confirmed }
    public var version: Int = 1
    public var stage: Stage
    public var confirmation: PairingConfirmation
    public var proof: String

    public static func signingInput(stage: Stage, confirmation: PairingConfirmation) -> Data {
        let fields = ["plink-consent-v1", stage.rawValue, confirmation.deviceId, confirmation.deviceName,
                      confirmation.platform, confirmation.endpoint, confirmation.publicKey,
                      confirmation.targetDeviceId, confirmation.offerNonce, confirmation.sessionId,
                      String(confirmation.protocolVersion)]
        return Data(fields.map { "\($0.utf8.count):\($0)" }.joined().utf8)
    }

    public static func make(stage: Stage, confirmation: PairingConfirmation, sessionKey: Data) throws -> Self {
        guard sessionKey.count == 32 else { throw PairingPayloadError.invalidPayload }
        let code = HMAC<SHA256>.authenticationCode(for: signingInput(stage: stage, confirmation: confirmation), using: SymmetricKey(data: sessionKey))
        return Self(stage: stage, confirmation: confirmation, proof: Data(code).base64EncodedString())
    }

    public func verified(using sessionKey: Data) throws -> Bool {
        guard version == 1, sessionKey.count == 32, let signature = Data(base64Encoded: proof) else { return false }
        return HMAC<SHA256>.isValidAuthenticationCode(signature, authenticating: Self.signingInput(stage: stage, confirmation: confirmation), using: SymmetricKey(data: sessionKey))
    }

    public func encode() throws -> String {
        let data = try JSONEncoder().encode(self)
        guard data.count <= 16_384 else { throw PairingPayloadError.invalidPayload }
        return String(decoding: data, as: UTF8.self)
    }

    public static func decode(_ payload: String) throws -> Self {
        guard payload.utf8.count <= 16_384 else { throw PairingPayloadError.invalidPayload }
        let value = try JSONDecoder().decode(Self.self, from: Data(payload.utf8))
        guard value.version == 1 else { throw PairingPayloadError.invalidPayload }
        return value
    }
}
